spec/integration/application/filebucket_spec.rb

```diff
@@ -70,10 +70,114 @@
         filebucket.command_line.args << 'fac251367c9e083c6b1f0f3181'
         filebucket.run
       }.to output(a_string_matching(
-        %r{Debug: HTTP GET https:\/\/127.0.0.1:#{port}\/puppet\/v3\/file_bucket_file\/md5\/fac251367c9e083c6b1f0f3181 returned 200 OK}
+        %r{Debug: HTTP GET https:\/\/127.0.0.1:#{port}\/puppet\/v3\/file_bucket_file\/md5\/fac251367c9e083c6b1f0f3181\?environment\=production returned 200 OK}
       ).and matching(
         %r{something to store}
        )).to_stdout
     end
   end
+
+  context 'diff', unless: Puppet::Util::Platform.windows? || Puppet::Util::Platform.jruby? do
+    context 'using a remote bucket' do
+      it 'outputs a diff between a local and remote file' do
+        File.binwrite(backup_file, "bar\nbaz")
+
+        get_handler = -> (req, res) {
+          res['Content-Type'] = 'application/octet-stream'
+          res.body = 'foo'
+        }
+
+        server.start_server(mounts: {filebucket: get_handler}) do |port|
+          Puppet[:masterport] = port
+          expect {
+            filebucket.command_line.args += ['diff', 'fac251367c9e083c6b1f0f3181', backup_file, '--remote']
+            filebucket.run
+          }.to output(a_string_matching(
+            /[-<] ?foo/
+          ).and matching(
+            /[+>] ?bar/
+          ).and matching(
+            /[+>] ?baz/
+          )).to_stdout
+        end
+      end
+
+      it 'outputs a diff between two remote files' do
+        get_handler = -> (req, res) {
+          res['Content-Type'] = 'application/octet-stream'
+          res.body = "--- /opt/puppetlabs/server/data/puppetserver/bucket/d/3/b/0/7/3/8/4/d3b07384d113edec49eaa6238ad5ff00/contents\t2020-04-06 21:25:24.892367570 +0000\n+++ /opt/puppetlabs/server/data/puppetserver/bucket/9/9/b/9/9/9/2/0/99b999207e287afffc86c053e5693247/contents\t2020-04-06 21:26:13.603398063 +0000\n@@ -1 +1,2 @@\n-foo\n+bar\n+baz\n"
```

Comment:
nit, to improve readability, I usually use [squiggly heredoc](https://www.rubyguides.com/2018/11/ruby-heredoc/) for multiline text, which strips leading whitespace:

```ruby
res.body = <<~END
  --- /opt/puppetlabs/server/data/puppetserver/bucket/d/3/b/0/7/3/8/4/d3b07384d113edec49eaa6238ad5ff00/contents\t2020-04-06 21:25:24.892367570 +0000
  +++ /opt/puppetlabs/server/data/puppetserver/bucket/9/9/b/9/9/9/2/0/99b999207e287afffc86c053e5693247/contents\t2020-04-06 21:26:13.603398063 +0000
  ...
END

Reply:
done